### PR TITLE
feat: allow sampling in eta

### PR DESCRIPTION
**Pull request by asalzburger**
This mini-PR adds a new option `--gen-sample-eta` to the `ParametricParticleGenerator` that allows to draw a flat eta distribution. This can be useful for detector performance classification, to populate the eta-phi plane uniquely. It can also be favourable for the material mapping to have a flat eta distribution.

The following shows the `eta` distributions of generated particles:
 * blue: standard sampling in `cos(theta)`
 * red: flat `eta` sampling with the `--gen-sample-eta` switch opted on
 
Screenshot 2021-09-23 at 14 37 25

### CODE REVIEW

**Comment by asalzburger** [comment]
Hey @mesmerisa - since the (new) material mapping uses the particle generation now, this might solve your problem - can you give it a try ?

**Comment by mesmerisa** [comment]
> Hey @mesmerisa - since the (new) material mapping uses the particle generation now, this might solve your problem - can you give it a try ?

Hi @asalzburger! Thank you, I will try it and let you know!

**Comment by asalzburger** [comment]
> > Hey @mesmerisa - since the (new) material mapping uses the particle generation now, this might solve your problem - can you give it a try ?
> 
> Hi @asalzburger! Thank you, I will try it and let you know!

Cool, if you can also review it, that would be great - there's really nothing special ... it just allows optional to draw the direction from eta and not cos(theta).

File: Examples/Algorithms/Generators/ActsExamples/Generators/ParametricParticleGenerator.hpp [hunk 1]
```diff
@@ -35,8 +35,14 @@ class ParametricParticleGenerator : public EventGenerator::ParticlesGenerator {
...
     /// This intentionally uses theta instead of eta so it can represent the
     /// full direction space with finite values.
+    ///
+    /// @note This is the standard generation, for detector performance
+    /// classification, where a flat distribution in eta can be useful,
+    /// this can be set by the forceEtaGeneration flag;
+    ///
     double thetaMin = 0.0;
     double thetaMax = M_PI;
+    bool forceEtaGeneration = false;
```
Comment (paulgessinger) [inline Examples/Algorithms/Generators/ActsExamples/Generators/ParametricParticleGenerator.hpp:45]: Can we also call this `sampleEta` or `uniformEta`?

Comment (asalzburger) [reply Examples/Algorithms/Generators/ActsExamples/Generators/ParametricParticleGenerator.hpp:45]: Absolutely, will change.

**Comment by mesmerisa** [comment]
> Cool, if you can also review it, that would be great - there's really nothing special ... it just allows optional to draw the direction from eta and not cos(theta).

Sure, I can also review it.
I just cloned your repo, switched to the branch and ran ActsExamplesMaterialRecordingGdml with the options --gen-eta 2:4 --gen-sample-eta and my geometry, and I still see only particles with momentum (0,0,1) in the resulting geant4_material_tracks.root. 

I also see the same without setting this option, in contrary to @Corentin-Allaire, who saw the same behaviour when running ActsExampleMaterialRecordingDD4hep for the open data detector but only when using "--gen-eta".
Because of that, I am wondering, could the problem lie somewhere else?

For example, I think that within the G4ParticleGun, if the particle definition is not properly set, then the particle is initialised with unit momentum. 

Also, I am using the docker image ghcr.io/acts-project/centos7-lcg98python3-gcc10 to run ACTS, in case that somehow could make a difference.

**Review by mesmerisa — APPROVED** [review]
Ok, I can confirm that the `--gen-sample-eta` option works as expected.

The problem I had with the momentum direction came from the energy conversion: `auto mom4 = part.fourMomentum() * convertEnergy;` in the SimParticleTranslation.cpp. If I remove the conversion (i.e. `auto mom4 = part.fourMomentum();`) then the output in the root file looks as expected.  Could this be due to different geant4 versions or something? (as mentioned in a previous comment, I am using one of the docker images to run acts).

File: Examples/Algorithms/Generators/ActsExamples/Generators/ParametricParticleGenerator.hpp (hunk 1 above)
Comment (asalzburger) [reply Examples/Algorithms/Generators/ActsExamples/Generators/ParametricParticleGenerator.hpp]: changed.

**Comment by asalzburger** [comment]
This was approved by @mesmerisa - merging it.

**Pull request merged**